File: imageSlicer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <math.h>
#include <climits>
#include <list>
#include "mongo/client/dbclient.h"
#include "CycleTimer.h"

#include "imageSlicer.h"

using namespace std;
using namespace Magick;
using namespace mongo;

ImageSlicer::ImageSlicer (string src, int n, int cSize) {
  imgsrc = src;
  numSlices = n;
  cutSize = cSize;
  slices = vector< vector<Image> > (n);
  for (int x = 0; x < n; x++) {
    slices[x] = vector<Image> (n);
  }
  sourceImage = Image(src);
  if (sourceImage.columns() != 612 || sourceImage.rows() != 612) {
    printf("Trying to slice an image that is not 612 x 612 at %s\n", src.c_str());
    exit(1);
  }
}

// slice the input image into a grid of numSlices x numSlices
// and puts these subimages into an array for use later.
void ImageSlicer::slice() {
  Image img = sourceImage;
  int width = img.columns();
  int height = img.rows(); // size of instagram photo

  if (numSlices == 1) {
    Image piece = img;
    slices[0][0] = img;
    return;
  }

  int subwidth = width / numSlices;
  int subheight = height / numSlices;
  int count = 0;
  for (int x = 0; x < numSlices; x++) {
    for (int y = 0; y < numSlices; y++) {
      Image piece = img;
      piece.crop(Geometry(subwidth, subheight, x * subwidth, y * subheight));
      // piece.write(sstr.str());
      slices[y][x] = piece;
      count++;
    }
  }
  return;
}

// get a list of RGB values from the input image. This takes each slice
// and cuts it into a 3x3 grid and finds the average RGB value in each
// of the grid sections. Thus each sub image is represented by 9 RGB values
void ImageSlicer::calculateRGBValues() {
  int subwidth = 612 / numSlices;
  int subheight = 612 / numSlices;

  int average_subwidth = subwidth / cutSize;
  int average_subheight = subheight / cutSize;

  for (int x = 0; x < numSlices; x++) {
    for (int y = 0; y < numSlices; y++) {
      Image currentSlice = slices[x][y];
      Pixels view(currentSlice);
      PixelPacket *pixels = view.get(0, 0, currentSlice.columns(), currentSlice.rows());
      for (int i = 0; i < cutSize; i++) {
        for (int j = 0; j < cutSize; j++) {
          int red = 0;
          int green = 0;
          int blue = 0;
          for (int k = 0; k < average_subwidth; k++) {
            for (int l = 0; l < average_subheight; l++) {
              int curX = (i * average_subwidth) + k;
              int curY = (j * average_subheight) + l;
              int pixelLoc = curX + (curY * subwidth);
              ColorRGB pixel = ColorRGB(pixels[pixelLoc]);
              red += pixel.red() * 255;
              green += pixel.green() * 255;
              blue += pixel.blue() * 255;
            }
          }
          int numIters = average_subwidth * average_subheight;
          red /= numIters;
          blue /= numIters;
          green /= numIters;
          RGB rgb;
          rgb.red = red;
          rgb.green = green;
          rgb.blue = blue;
          // printf("Red: %d, Green: %d, Blue: %d\n", red, green, blue);
          averages.push_back(rgb);
        }
      }
    }
  }
}

vector< vector< Image > > ImageSlicer::getSlices() {
  slice();
  return slices;
}

vector<RGB> ImageSlicer::getAverages() {
  slice();
  calculateRGBValues();
  return averages;
}

File: imageSlicer.py
import os
import sys
from PIL import Image
from pymongo import MongoClient

# raised when num_slices is not a divisor of instagram_wh
class InstagramSizeException(Exception):
  pass

class ImageSlicer:
  """
  Image slicer

  Attributes:
    imgsrc     -- string specifiying the location of the image to be sliced
    num_slices -- number of slices to cut the image into. Ex: 12 = (12 x 12)
    saveloc    -- location to save the resulting slices 
  """

  # default width and height of instagram photos.
  instagram_wh = 612

  def __init__(self, imgsrc, num_slices=12):
    self.imgsrc = imgsrc
    self.num_slices = num_slices
    self.slices = []
    self.rgbs = []
    self.averages = []
    if ImageSlicer.instagram_wh % num_slices != 0:
      raise InstagramSizeException("num_slices is not a divisor of 612")
    self.client = MongoClient()

  def get_averages(self):
    if (self.averages == []):
      self.average_RGB()
    return self.averages

  # slice image into pieces of equal size defined by num_slices
  def slice(self):
    if (self.slices != []):
      return self.slices
    img = Image.open(self.imgsrc)
    width, height = img.size
    slices = [[None for w in range(self.num_slices)] for h in range(self.num_slices)]
    if (width != ImageSlicer.instagram_wh or 
       height != ImageSlicer.instagram_wh):
      raise InstagramSizeException("Photo is not 612 x 612")

    subwidth = ImageSlicer.instagram_wh / self.num_slices
    subheight = ImageSlicer.instagram_wh / self.num_slices
    for x in xrange(self.num_slices):
      for y in xrange(self.num_slices):
        box = [
          x * subwidth,
          y * subheight,
          (x + 1) * subwidth,
          (y + 1) * subheight
        ]
        piece = img.crop(box)
        slices[y][x] = (piece)
    self.slices = slices
    return slices

  def get_RGB_values(self):
    if self.slices is None or self.slices == []:
      self.slice()
    self.rgbs = map(lambda y: map(lambda x: list(x.getdata()), y), self.slices)
    return self.rgbs
    
  def average_RGB(self):

    cut_size = 3
    # def reduce_average((r1,g1,b1), (r2,g2,b2)):
    #   return ((r1 + r2) / 2, (g1 + g2) / 2, (b1 + b2) / 2)

    def reduce_average(l):
      rf = 0
      gf = 0
      bf = 0
      for (r,g,b) in l:
        rf += r
        gf += g
        bf += b
      num = len(l)
      return (rf / num, gf / num, bf / num)

    subpieces = []
    if (self.rgbs == []):
      self.get_RGB_values()
    subwidth = ImageSlicer.instagram_wh / self.num_slices
    subheight = ImageSlicer.instagram_wh / self.num_slices

    average_subwidth = subwidth / cut_size
    average_subheight = subheight / cut_size

    averages = []
    for x in xrange(self.num_slices):
      for y in xrange(self.num_slices):
        current_slice = self.slices[x][y]
        for i in xrange(cut_size):
          for j in xrange(cut_size):
            sub_averages = []
            cropped = current_slice.crop((average_subwidth * j,
                                          average_subheight * i,
                                          average_subwidth * j + average_subwidth,
                                          average_subheight * i + average_subheight))
            reduced = reduce_average(list(cropped.getdata()))
            # print reduced
            sub_averages.append(reduced)
            averages.append(sub_averages)
    self.averages = averages
    
    flattened = []
    block_size = self.num_slices*cut_size**2
    h = 0
    while (h != self.num_slices):
      for k in range(block_size*h, block_size*(h+1)):
        if (k % (cut_size**2) == 0 or k % (cut_size**2) == 1 or k % (cut_size**2) == 2):
          flattened.append(averages[k])
      
      for m in range(block_size*h, block_size*(h+1)):
        if (m % (cut_size**2) == 3 or m % (cut_size**2) == 4 or m % (cut_size**2) == 5) :
          flattened.append(averages[m])
      
      for n in range(block_size*h, block_size*(h+1)):
        if (n % (cut_size**2) == 6 or n % (cut_size**2) == 7 or n % (cut_size**2) == 8) :
          flattened.append(averages[n])          
      h+=1
    #print flattened
    return [item for sublist in flattened for item in sublist]


# slicer = ImageSlicer("images/318960010740589992_182714056.jpg", 1)

# # # slicer.average_RGB()
# # # exit(0)

# img.putdata(averages)
# img.save("./test3.jpg")
# # Image manip
# img = Image.open(path)
# # get a list of all rgb values in the image
# list(img.getdata())













